In [26]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression

In [14]:
file  = pd.read_excel(r'C:\Users\Brian Lin\Desktop\UCI\3rd quarter\Applied Forecast Modeling\NFL event\newdata.xlsx')
test = pd.read_excel(r'C:\Users\Brian Lin\Desktop\UCI\3rd quarter\Applied Forecast Modeling\NFL event\NFL Data1.xlsx')

In [194]:
list(file)

['Team1',
 'Team2',
 'Team1_Score',
 'Team2_Score',
 'Team1_Passing',
 'Team1_Rushing',
 'Team1_Turnovers',
 'Team2_Passing',
 'Team2_Rushing',
 'Team2_Turnovers',
 'Team1_win',
 'team1_offence_rushing',
 'team1_offence_passing ',
 'team1_offence_overall',
 'team1_defence_rushing',
 'team1_defence_passing',
 'team1_defence_overall ',
 'team2_defence_rushing',
 'team2_defence_passing',
 'team2_defence_overall',
 'team2_offence_rushing',
 'team2_offence_passing',
 'team2_offence_overall']

In [192]:
filename =list(file)[11:]
testnew = test.iloc[:,1:]
testnew.columns = filename 


In [219]:
test

Team1  offence1_RUSHING  offence1_PASSING  offence1_OVERALL  \
0  Philadelphia Eagles             147.6             256.7             389.1   

   defence1_RUSHING  defence1_PASSING  defence1_OVERALL  offence_RUSHING  \
0             121.6             208.7             301.5            115.9   

   offence_PASSING  offence_OVERALL  defence_RUSHING  defence_PASSING  \
0            308.8            413.6            107.2            243.8   

   defence_OVERALL  
0            328.2

In [107]:
list(file)

file.head()

Team1                 Team2  Team1_Score  Team2_Score  \
0  Philadelphia Eagles     Minnesota Vikings           24            7   
1  Philadelphia Eagles         Chicago Bears           25           20   
2   Kansas City Chiefs  Los Angeles Chargers           27           24   
3   Kansas City Chiefs  Los Angeles Chargers           30           27   
4  Philadelphia Eagles  Jacksonville Jaguars           29           21   

   Team1_Passing  Team1_Rushing  Team1_Turnovers  Team2_Passing  \
0            323            163                1            202   
1            309            112                3             91   
2            226             93                0            326   
3            322            163                1            250   
4            191            210                1            148   

   Team2_Rushing  Team2_Turnovers  ...  team1_offence_overall  \
0             62                3  ...                  389.1   
1            157                1  ...                  389.1   
2             75                1  ...                  413.6   
3            115                2  ...                  413.6   
4             71                5  ...                  389.1   

   team1_defence_rushing  team1_defence_passing  team1_defence_overall   \
0                  121.6                  208.7                   301.5   
1                  121.6                  208.7                   301.5   
2                  107.2                  243.8                   328.2   
3                  107.2                  243.8                   328.2   
4                  121.6                  208.7                   301.5   

   team2_defence_rushing  team2_defence_passing  team2_defence_overall  \
0                  123.1                  279.8                  388.7   
1                  157.3                  225.9                  375.9   
2                  145.8                  217.2                  346.1   
3                  145.8                  217.2                  346.1   
4                  114.8                  250.5                  353.3   

   team2_offence_rushing  team2_offence_passing  team2_offence_overall  
0                   97.7                  283.4                  361.5  
1                  177.3                  152.8                  307.8  
2                   89.6                  281.8                  359.3  
3                   89.6                  281.8                  359.3  
4                  124.5                  244.0                  357.4  

[5 rows x 23 columns]

***Linear regression***

In [39]:
#team 1 score model
X_linear = file.iloc[:,11:]
y_linear = file.iloc[:,2]

In [257]:
reg = LinearRegression().fit(X_linear, y_linear)
reg.score(X_linear, y_linear)


MSE_linear = mean_squared_error(list(y_linear), reg.predict(X_linear))
 
RMSE_linear = math.sqrt(MSE)

print('prediction for team 1 score: ',reg.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(reg.score(X_linear, y_linear),2))
print('MAPE : ',round(mape(list(y_linear), reg.predict(X_linear)),2))
print('RSME : ',round(RMSE_linear,2))
print('MSE : ',round(MSE_linear,2))


prediction for team 1 score:  33.03125
accuracy rate:  0.13
MAPE :  0.24
RSME :  9.19
MSE :  56.84


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [284]:
#team 2 score model
X_linear_2 = file.iloc[:,11:]
y_linear_2 = file.iloc[:,3]
reg1 = LinearRegression().fit(X_linear_2, y_linear_2)

MSE_linear1 = mean_squared_error(list(y_linear_2), reg1.predict(X_linear_2))
 
RMSE_linear1 = math.sqrt(MSE)

print('prediction for team 2 score: ',reg1.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(reg1.score(X_linear_2 , y_linear_2),2))
print('MAPE : ',round(mape(list(y_linear_2), reg1.predict(X_linear_2)),2))
print('RSME : ',round(RMSE_linear1,2))
print('MSE : ',round(MSE_linear1,2))


prediction for team 2 score:  12.203125
accuracy rate:  0.22
MAPE :  0.37
RSME :  9.19
MSE :  50.38


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


***logistic regression***

In [262]:
X_logic = file.iloc[:,11:]
y_logic = file.iloc[:,10]
clf_log = LogisticRegression(random_state=0).fit(X_logic, y_logic)



MSE_logic = mean_squared_error(list(y_neu2), clf_new1.predict(X_neu2))
 
RMSE_logic = math.sqrt(MSE1)

print('prediction for team 2 score: ',clf_log.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(clf_log .score(X_logic, y_logic),2))
print('recall score: ',round(recall_score(list(y_logic), clf_log.predict(X_logic), average='macro'),2))
print('MAPE : ',round(mape(list(y_logic), clf_log.predict(X_logic)),2))
print('RSME : ',round(MSE_logic,2))
print('MSE : ',round(RMSE_logic,2))

prediction for team 2 score:  1
accuracy rate:  0.89
recall score:  0.73
MAPE :  nan
RSME :  81.35
MSE :  9.02


C:\Users\Public\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offen

In [112]:
#wining rate: philly
clf_log.predict_proba(test.iloc[:,1:])

C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


array([[3.08366612e-06, 9.99996916e-01]])

***neural network***

In [240]:
def mape(a,b):
    APE = []
  
    # Iterate over the list values
    for day in range(len(a)):

        # Calculate percentage error
        per_err = (a[day] - b[day]) / a[day]

        # Take absolute value of
        # the percentage error (APE)
        per_err = abs(per_err)

        # Append it to the APE list
        APE.append(per_err)
    MAPE = sum(APE)/len(APE)
    return MAPE

In [251]:
#team1 score: philly
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import recall_score
from sklearn.metrics import mean_squared_error
import math

X_neu1 = file.iloc[:,11:]
y_neu1 = file.iloc[:,2]

clf_new1 = MLPClassifier(solver='lbfgs', alpha=1e-5,
                    hidden_layer_sizes=(10, 2), random_state=1)
clf_new1.fit(X_neu1,y_neu1)
MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1,
              solver='lbfgs')

MSE = mean_squared_error(list(y_neu1), clf_new1.predict(X_neu1))
 
RMSE = math.sqrt(MSE)

print('prediction for team 1 score: ',clf_new1.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(clf_new1.score(X_neu1,y_neu1),2))
print('recall score: ',round(recall_score(list(y_neu1), clf_new1.predict(X_neu1), average='macro'),2))
print('MAPE : ',round(mape(list(y_neu1), clf_new1.predict(X_neu1)),2))
print('RSME : ',round(RMSE,2))
print('MSE : ',round(MSE,2))

prediction for team 1 score:  24
accuracy rate:  0.11
recall score:  0.05
MAPE :  0.24
RSME :  9.19
MSE :  84.41


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [252]:
#team2 score: kansas
from sklearn.neural_network import MLPClassifier
X_neu2 = file.iloc[:,11:]
y_neu2 = file.iloc[:,3]


clf_1_new2= MLPClassifier(solver='lbfgs', alpha=1e-5,
                    hidden_layer_sizes=(10, 2), random_state=1)
clf_1_new2.fit(X_neu2,y_neu2)
MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1,
              solver='lbfgs')


MSE1 = mean_squared_error(list(y_neu2), clf_new1.predict(X_neu2))
 
RMSE1 = math.sqrt(MSE1)

print('prediction for team 2 score: ',clf_1_new2.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(clf_new1.score(X_neu2,y_neu2),2))
print('recall score: ',round(recall_score(list(y_neu2), clf_1_new2.predict(X_neu2), average='macro'),2))
print('MAPE : ',round(mape(list(y_neu2), clf_1_new2.predict(X_neu2)),2))
print('RSME : ',round(RMSE1,2))
print('MSE : ',round(MSE1,2))

prediction for team 2 score:  17
accuracy rate:  0.11
recall score:  0.06
MAPE :  0.4
RSME :  9.02
MSE :  81.35


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [282]:
#team1 win
from sklearn.neural_network import MLPClassifier
X_neu_win = file.iloc[:,11:]
y_neu_win = file.iloc[:,10]


clf_1_new2_win= MLPClassifier(solver='lbfgs', alpha=1e-5,
                    hidden_layer_sizes=(10, 2), random_state=1)
clf_1_new2_win.fit(X_neu_win,y_neu_win)
MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1,
              solver='lbfgs')


MSE1_win = mean_squared_error(list(y_neu_win), clf_1_new2_win.predict(X_neu_win))
 
RMSE1_win = math.sqrt(MSE1_win)

print('prediction for team 1 win: ',clf_1_new2_win.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(clf_1_new2_win.score(X_neu_win,y_neu_win),2))
print('recall score: ',round(recall_score(list(y_neu_win), clf_1_new2_win.predict(X_neu_win), average='macro'),2))
print('MAPE : ',round(mape(list(y_neu_win), clf_1_new2_win.predict(X_neu_win)),2))
print('RSME : ',round(RMSE1_win,2))
print('MSE : ',round(MSE1_win,2))

prediction for team 1 win:  1
accuracy rate:  0.84
recall score:  0.5
MAPE :  inf
RSME :  0.4
MSE :  0.16


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)
C:\Users\Brian Lin\AppData\Local\Temp\ipykernel_20660\2346897411.py:8: RuntimeWarning: divide by zero encountered in longlong_scalars
  per_err = (a[day] - b[day]) / a[day]


***random forest***

In [209]:
from sklearn.ensemble import RandomForestRegressor
import numpy as np
from sklearn.metrics import mean_absolute_error

#team 1 score : philly
X_rf = file.iloc[:,11:]
y_rf = file.iloc[:,2]
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)

RandomForestRegressor(n_estimators=1000, random_state=42)



In [267]:
predictions = rf.predict(test.iloc[:,1:])
print(predictions)
errors1 = abs(predictions - y_rf)
print('Mean Absolute Error:', round(np.mean(errors1), 2), 'degrees.')

MSE_rf = mean_squared_error(list(y_neu2), rf.predict(X_neu2))
 
RMSE_rf = math.sqrt(MSE1)
errors = abs(predictions - y_rf)

print('prediction for team 1 score: ',rf.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(rf.score(X_rf, y_rf),2))
print('Mean Absolute Error: ',round(np.mean(errors), 2))
# print('recall score: ',round(recall_score(list(y_rf), rf.predict(X_rf), average='macro'),2))
print('MAPE : ',round(mape(list(y_rf), rf.predict(X_rf)),2))
print('RSME : ',round(MSE_rf,2))
print('MSE : ',round(RMSE_rf,2))

C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)
C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_

[26.24613095]
Mean Absolute Error: 6.32 degrees.
prediction for team 1 score:  26.24613095238095
accuracy rate:  0.68
Mean Absolute Error:  6.32
MAPE :  0.13
RSME :  157.02
MSE :  9.02


In [268]:
#team 2 score : kansas
X_rf1 = file.iloc[:,11:]
y_rf1 = file.iloc[:,3]
rf1 = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf1.fit(X_rf1, y_rf1)
predictions1 = rf1.predict(test.iloc[:,1:])
print("prediction of team 2 ",predictions)
errors1 = abs(predictions1 - y_rf1)



MSE_rf1= mean_squared_error(list(y_neu2), rf.predict(X_neu2))
 
RMSE_rf1= math.sqrt(MSE1)
errors = abs(predictions - y_rf1)

print('prediction for team 2 score: ',rf.predict(test.iloc[:,1:])[0])
print('accuracy rate: ',round(rf1.score(X_rf1, y_rf1),2))
print('Mean Absolute Error:', round(np.mean(errors1), 2), 'degrees.')
# print('recall score: ',round(recall_score(list(y_rf), rf.predict(X_rf), average='macro'),2))
print('MAPE : ',round(mape(list(y_rf1), rf.predict(X_rf1)),2))
print('RSME : ',round(MSE_rf1,2))
print('MSE : ',round(RMSE_rf1,2))

prediction of team 2  [26.24613095]
prediction for team 2 score:  26.24613095238095
accuracy rate:  0.55
Mean Absolute Error: 6.51 degrees.
MAPE :  0.81
RSME :  157.02
MSE :  9.02


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)
C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_

In [277]:
#team 1 win
X_rf_win = file.iloc[:,11:]
y_rf_win = file.iloc[:,10]
rf_win = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf_win.fit(X_rf_win, y_rf_win)
predictions = rf_win.predict(test.iloc[:,1:])
print("prediction of team 1 ",predictions)
errors_win = abs(predictions - y_rf_win)
print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

MSE_rf_win= mean_squared_error(list(y_rf_win), rf.predict(X_rf_win))
 
RMSE_rf_win= math.sqrt(MSE_rf_win)


print('prediction for team 1 wining: ',rf_win.predict(test.iloc[:,1:]))
print('accuracy rate: ',round(rf_win.score(X_rf_win, y_rf_win),2))
print('Mean Absolute Error:', round(np.mean(errors_win), 2), 'degrees.')
# print('recall score: ',round(recall_score(list(y_rf_win), rf_win.predict(X_rf_win), average='macro'),2))
# print('MAPE : ',round(mape(list(y_rf_win), rf_win.predict(X_rf_win)),2))
print('RSME : ',round(MSE_rf_win,2))
print('MSE : ',round(RMSE_rf_win,2))

prediction of team 1  [0.989]
Mean Absolute Error: 8.57 degrees.
prediction for team 1 wining:  [0.989]
accuracy rate:  0.62
Mean Absolute Error: 0.17 degrees.
RSME :  790.21
MSE :  28.11


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)
C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_

***knn***

In [288]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error
from math import sqrt
#team 1 score
X_knn = file.iloc[:,11:]
y_knn = file.iloc[:,2]
knn_model = KNeighborsRegressor(n_neighbors=3)
knn_model.fit(X_knn, y_knn)
train_preds = knn_model.predict(X_knn)

mse = mean_squared_error(train_preds, y_knn)
rmse = sqrt(mse)
preds = knn_model.predict(test.iloc[:,1:])

errors_knn = abs(train_preds - y_knn)


MSE_rf_win= mean_squared_error(list(y_rf_win), rf.predict(X_rf_win))
 
RMSE_rf_win= math.sqrt(MSE_rf_win)


print('prediction for team 1 score: ',preds)
print('accuracy rate: ',round(knn_model.score(X_knn, y_knn),2))
print('Mean Absolute Error:', round(np.mean(errors_win), 2), 'degrees.')
#print('recall score: ',round(recall_score(list(y_knn),knn_model.predict(X_knn), average='macro'),2))
print('MAPE : ',round(mape(list(y_knn), knn_model.predict(X_knn)),2))
print('RSME : ',round(rmse,2))
print('MSE : ',round(mse,2))

prediction for team 1 score:  [30.33333333]
accuracy rate:  0.24
Mean Absolute Error: 0.17 degrees.
MAPE :  0.22
RSME :  7.05
MSE :  49.69


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [289]:
#team 1 score
X_knn1 = file.iloc[:,11:]
y_knn1 = file.iloc[:,3]
knn_model1 = KNeighborsRegressor(n_neighbors=3)
knn_model1.fit(X_knn1, y_knn1)
train_preds1 = knn_model1.predict(X_knn1)

mse1 = mean_squared_error(train_preds, y_knn1)
rmse1 = sqrt(mse)

preds1 = knn_model1.predict(test.iloc[:,1:])


errors_knn = abs(train_preds1 - y_knn)


MSE_rf_win= mean_squared_error(list(y_rf_win), rf.predict(X_rf_win))
 
RMSE_rf_win= math.sqrt(MSE_rf_win)


print('prediction for team 2 score: ',preds1)
print('accuracy rate: ',round(knn_model1.score(X_knn1, y_knn1),2))
print('Mean Absolute Error:', round(np.mean(errors_win), 2), 'degrees.')
#print('recall score: ',round(recall_score(list(y_knn),knn_model.predict(X_knn), average='macro'),2))
print('MAPE : ',round(mape(list(y_knn), knn_model1.predict(X_knn1)),2))
print('RSME : ',round(rmse1,2))
print('MSE : ',round(mse1,2))

prediction for team 2 score:  [21.]
accuracy rate:  0.3
Mean Absolute Error: 0.17 degrees.
MAPE :  0.31
RSME :  7.05
MSE :  146.74


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


***Naive bayes***

In [160]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix,accuracy_score
#team 1 score
X_naive1 = file.iloc[:,11:]
y_naive1 = file.iloc[:,2]

naive_classifier1 = GaussianNB()
naive_classifier1.fit(X_naive1, y_naive1)
y_pred_naive1  =  naive_classifier1.predict(X_naive1)
cm1 = confusion_matrix(y_naive1, y_pred_naive1)
ac1 = accuracy_score(y_naive1,y_pred_naive1)
print("team 1 : ",ac1)
print(naive_classifier1.predict(test.iloc[:,1:]))

team 1 :  0.5405405405405406
[29]


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [159]:
#team 2 score
X_naive2 = file.iloc[:,11:]
y_naive2 = file.iloc[:,3]

naive_classifier2 = GaussianNB()
naive_classifier2.fit(X_naive2, y_naive2)
y_pred_naive2  =  naive_classifier2.predict(X_naive2)
cm2 = confusion_matrix(y_naive2, y_pred_naive2)
ac2 = accuracy_score(y_naive2,y_pred_naive2)
print("team 2 : ",ac2)
print(naive_classifier2.predict(test.iloc[:,1:]))

team 2 :  0.40540540540540543
[7]


C:\Users\Public\anaconda3\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- defence1_OVERALL
- defence1_PASSING
- defence1_RUSHING
- defence_OVERALL
- defence_PASSING
- ...
Feature names seen at fit time, yet now missing:
- team1_defence_overall 
- team1_defence_passing
- team1_defence_rushing
- team1_offence_overall
- team1_offence_passing 
- ...

  warnings.warn(message, FutureWarning)


In [197]:
print(list(file.iloc[:,11:]))
print(list(testnew))

['team1_offence_rushing', 'team1_offence_passing ', 'team1_offence_overall', 'team1_defence_rushing', 'team1_defence_passing', 'team1_defence_overall ', 'team2_defence_rushing', 'team2_defence_passing', 'team2_defence_overall', 'team2_offence_rushing', 'team2_offence_passing', 'team2_offence_overall']
['team1_offence_rushing', 'team1_offence_passing ', 'team1_offence_overall', 'team1_defence_rushing', 'team1_defence_passing', 'team1_defence_overall ', 'team2_defence_rushing', 'team2_defence_passing', 'team2_defence_overall', 'team2_offence_rushing', 'team2_offence_passing', 'team2_offence_overall']


In [211]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error
import pandas as pd
import numpy as np

#team 1 score
X_xgb = file.iloc[:,11:]
y_xgb = file.iloc[:,2]

data_dmatrix = xgb.DMatrix(data=X_xgb,label=y_xgb)
xg_reg = xgb.XGBRegressor(objective ='reg:linear', colsample_bytree = 0.3, learning_rate = 0.1,
                max_depth = 5, alpha = 10, n_estimators = 10)
xg_reg.fit(X_xgb,y_xgb)

preds = xg_reg.predict(X_xgb)
rmse = np.sqrt(mean_squared_error(y_xgb, preds))
print("RMSE: %f" % (rmse))
print(xg_reg.predict(testnew))

[16:22:41] WARNING: C:/buildkite-agent/builds/buildkite-windows-cpu-autoscaling-group-i-08de971ced8a8cdc6-1/xgboost/xgboost-ci-windows/src/objective/regression_obj.cu:213: reg:linear is now deprecated in favor of reg:squarederror.
RMSE: 12.851722
[18.36897]


In [215]:
#team 2 score
X_xgb1 = file.iloc[:,11:]
y_xgb1 = file.iloc[:,3]

data_dmatrix1 = xgb.DMatrix(data=X_xgb1,label=y_xgb1)
xg_reg1 = xgb.XGBRegressor(objective ='reg:linear', colsample_bytree = 0.3, learning_rate = 0.1,
                max_depth = 5, alpha = 10, n_estimators = 10)
xg_reg1.fit(X_xgb1,y_xgb1)

preds1 = xg_reg1.predict(X_xgb1)
rmse1 = np.sqrt(mean_squared_error(y_xgb1, preds1))
print("RMSE: %f" % (rmse1))
print(xg_reg1.predict(testnew))

[16:24:57] WARNING: C:/buildkite-agent/builds/buildkite-windows-cpu-autoscaling-group-i-08de971ced8a8cdc6-1/xgboost/xgboost-ci-windows/src/objective/regression_obj.cu:213: reg:linear is now deprecated in favor of reg:squarederror.
RMSE: 10.544454
[10.913336]


<font size="15">conclusion </font>




***Linear regression: 
<br>
<br>
R-square: team_1_score: .125
<br>
R-square: team_2_score: .218
<br>
score(philly v.s. kansas): 33:12***
***


***Logistic regression: 
<br>
<br>
R-square: 0.89
<br>
liklihood(philly win): 0.999997***
***


***Neural network: 
<br>R-square: team_1_score: .14
<br>
R-square: team_2_score: .11
<br>
R-square: Wining prob: 0.89
<br>
score(philly v.s. kansas): 24:17***
***

***random forest: 
<br>
<br>
R-square: team_1_score: .125
<br>
R-square: team_2_score: .218
<br>
score(philly v.s. kansas): 26:18
<br>
liklihood(philly win): 0.989***

***KNN: 
<br>
<br>
RMSE: team_1_score: 7.04
<br>
RMSE: team_2_score: 6.71
<br>
score(philly v.s. kansas): 30:20***


***Naive Bayes: 
<br>
<br>
Accuracy score: 
Team wining : 0.84<br>
team 1 score : 0.540<br>
team 2 score : 0.405
<br>
score(philly v.s. kansas): 21:7
<br>
liklihood(philly win): 1***


***XGBoost***


***RMSE: team_1_score: 12.85
<br>
RMSE: team_2_score: 10.54
<br>
score(philly v.s. kansas): 18:10***

***Guestimation:27:19, Philly wins***

**It's impossible to use simple model to predict the wining or loosing in a sport event. Each team will give everything they have and set a a special strategy in the final. Win or go home, that's how sport it is!**|